### ISSUE TITLE
Ability to peek messages based on status

### ISSUE THREAD

**Issue by SeanFeldman** [Azure/azure-service-bus#95]
Allow to query messages based on their state rather than receiving messages in all states and filtering on the client side.

From the [comment](https://github.com/Azure/azure-service-bus-dotnet/issues/343#issuecomment-334954288) this could be an option for the Premium Tier as that one uses customers dedicated MU pools.

**Comment by ChristianWolf42** [comment]
@SeanFeldman we put it on our backlog and consider it for our planning.

**Comment by SeanFeldman** [comment]
That would be great.

**Comment by SeanFeldman** [comment]
Was there any development on this? Thanks.

**Comment by nemakam** [comment]
@SeanFeldman 
This is in the backlog, but not a planned item. There are higher priority items that need more immediate attention. From the looks of it, it might take a while before we get to this feature.

**Comment by mcintyre94** [comment]
Sean commented on an SO question of mine where this would have helped (https://stackoverflow.com/questions/53710350/azure-service-bus-can-i-view-scheduled-messages/53713041#53713041)

Essentially we're using service bus as a lightweight task queue, with tasks able to be scheduled to run at a future time and a function app running as tasks become available. I'd like to be able to build a UI around this to show upcoming tasks without independently maintaining a list of all the tasks and when they're scheduled, and this would probably be the most straightforward way to do that.

**Comment by abhikt48** [comment]
It will be great if we can have this feature. This feature will be very useful from support prospective.

**Comment by qsdfplkj** [comment]
Wouldn't receiving them using the sequence number have the same effect as a purge.

**Comment by triynko** [comment]
We have a situation where there are like 3 million messages in a deadletter queue dating years back, because we don't always maintain them, knowing we can eventually just purge it.
This is a huge problem when an 'incident' occurs, and we need to replay a few recent deadletters. Assuming we want to replay them in a non-destructive way (leaving them in the DL queue), we have to peek through the entire DL queue from the beginning just to reach more recent records.  With standard tier's 256KB batch limit, It took us over 21 hours to read through messages. Of course, the VPN cut out at 21 hours when it was only 90% through inspecting messages, so now we have to retry the whole operation. Didn't keep a list of which sequence numbers were visited, for example. And, with the browsing order apparently being non-deterministic (I've read that it's "whichever broker happens to respond first?"), we can't just skip or read past the number we've already read. No idea how many messages are in each broker, and so on. It's a real pain.

With custom server-side filtering, at least we'd be able to locate the messages we're interested in (i.e. only ones with enqueue times during the incident time window.).

First, we could use multiple readers if we were actually removing messages as we read them, but we're not. We're peeking non-destructively; we want to leave the message there in case something goes wrong part way through the replay operation, as it did.

So, assuming peek-only mode, the reason we have to read with only a single thread is that if we were to use multiple QueueClients, that would result in each one re-reading all the same messages; no way to filter messages or coordinate amounts them, at least not out of the box. A workaround may be to craft starting sequence numbers with explicit partition bytes, such that we can run 16 QueueClients in parallel, each targetting a different partition, and stopping each one when it reaches the end of the partition. I'm assuming we can detect that when we receive messages with a different partition number than what we started with.

**Comment by LeszekKalibrate** [comment]
Is it implemented yet?

**Comment by EldertGrootenboer** [comment]
This has surfaced to the top of our backlog, so expect more information in the coming months, although we don't have specific timelines yet.

**Comment by OmarAlaniJBhifi** [comment]
Any updates on this feature guys ?

**Comment by AlphaCreative-Mike** [comment]
@EldertGrootenboer - Any updates?

**Comment by kominko** [comment]
@EldertGrootenboer any movement on this?

**Comment by EldertGrootenboer** [comment]
Thank you for your feedback on this item. We are currently doing active development on this feature, and expect to have more to share around its release in the next couple of months.

**Comment by jpeckham** [comment]
how are we doing on this? i have 8 scheduled messages that i have no idea when they will deliver and have no way to cancel them or modify their schedule.

some day i hope to know what they are.

**Comment by EldertGrootenboer** [comment]
We are currently doing active development on this feature, and expect to have more to share around its release in the next couple of months.

**Comment by Paizo** [comment]
Hello @EldertGrootenboer, any news on this feature?

**Comment by krispenner** [comment]
Hi @EldertGrootenboer it's been 14 months since your team started development on this - is there any timeline or detail to share?

**Comment by EldertGrootenboer** [comment]
A quick update on this thread. During development we ran into a situation, where we noticed that introducing this feature would bring a noticeable performance impact, and as such we are currently looking into ways to resolve this. As this would probably require changes to our architecture, this is currently on hold. We do understand that this feature is still sought after, however, we currently don't have any specific timelines to share. Once we have more information on this, we will come back and update this thread again. In the meanwhile, we do encourage you to keep sharing your scenarios.

**Comment by SeanFeldman** [comment]
Appreciate the transparency, @EldertGrootenboer.

**Comment by grujicd** [comment]
As you saw in this thread, if there is no official way - users will do it anyway on the client side. This also brings performance hit, you're not avoiding it, you're just not aware of it! Users are already reading millions of messages if there's no other way. Since performance penalty is already here, it should not be the reason to avoid this feature.

The main difference is that user has much worse experience since it's very inefficient to pull all these messages across network. At least it would be better for the user if filtering is done on the server end, and you'll have an option to improve it in the future and all the clients who use such feature will immediately get that improvement. 

One more thing - if you release this feature now, it will take many years for users to even notice it exists, and then to change their code to use it. So postponing this feature also postpones start of this long adoption period.

**Comment by sean-feldman-sh** [comment]
> As this would probably require changes to our architecture

@grujicd, let's be honest, there's no point in repetition of the request if the team needs to make some fundamental changes to enable the change. Lengthy petition won't speed it up.

**Comment by grujicd** [comment]
As I understood, "quick&dirty" implementation doesn't require any architecture changes, but has a performance impact. Architecture changes are only needed to avoid that hit.

My whole argument was that performance impact is already here. We're already doing it the inefficient way. It's just not visible to Service Bus team that such feature is already implemented by us users - they see it just as millions of reads and have no idea that we're doing it.

**Comment by SeanFeldman** [comment]
Reading between the lines is a skill 🙂 

I totally see the point you're trying to convey, but when there's a bigger fish to fry, the team's focus is likely on that BBQ party. That's all I can think of. [Believe me](https://github.com/Azure/azure-service-bus/issues/95#issue-263660526), I'm the last person not to want to see this feature, but I also acknowledge the reality.
